new-packages/ts-project/src/state.ts

```diff
@@ -0,0 +1,203 @@
+import type { Expression } from 'typescript';
+import { ExtractionContext, ExtractorStateConfig } from './types';
+import { getPropertyKey } from './utils';
+
+const isUndefined = (ts: typeof import('typescript'), prop: Expression) =>
+  ts.isIdentifier(prop) && ts.idText(prop) === 'undefined';
+
+const StateConfigDataWhitelistedProperties = new Set([
+  'key',
+  'type',
+  'history',
+  'metaEntries',
+  'entry',
+  'exit',
+  'invoke',
+  'initial',
+  'tags',
+  'description',
+]);
+
+const isStateNodeConfigProperty = (
+  key: string,
+): key is keyof ExtractorStateConfig['data'] =>
+  StateConfigDataWhitelistedProperties.has(key);
+
+export function extractState(
+  ctx: ExtractionContext,
+  ts: typeof import('typescript'),
+  state: Expression | undefined,
+  path: string[],
+): ExtractorStateConfig | undefined {
+  const result = {
+    states: [] as ExtractorStateConfig[],
+    data: {} as Partial<ExtractorStateConfig['data']>,
+  };
+
+  if (!state) {
+    return;
+  }
+
+  if (!ts.isObjectLiteralExpression(state)) {
+    ctx.errors.push({
+      type: 'state_unhandled',
+      node: state,
+    });
+    return;
+  }
+
+  for (const prop of state.properties) {
+    if (ts.isPropertyAssignment(prop)) {
+      const key = getPropertyKey(ctx, ts, prop);
+
+      /* state.states */
+      if (key === 'states' && ts.isObjectLiteralExpression(prop.initializer)) {
```

Comment:
This can easily be flattened~. We should be iterating over property assignments, try to get their keys and then switch/case through all known keys. This would allow you to drop `isStateNodeConfigProperty` and convert this if/elseif into a single switch/case